File: stacks/evaluate_reverse_polish_notation/eval_polish_notation_test.py
import pytest
from eval_polish_notation import eval_rpn


@pytest.mark.parametrize(
    "tokens, expected",
    [
        # Basic addition and multiplication: ((2 + 1) * 3) = 9
        (["2", "1", "+", "3", "*"], 9),
        # Division and addition: (4 + (13 / 5)) = 6
        (["4", "13", "5", "/", "+"], 6),
        # Complex expression: ((10 * (6 / ((9 + 3) * -11))) + 17) + 5
        (["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"], 22),
        # Truncation toward zero: (6 / -132) should be 0, not -1
        (["6", "-132", "/"], 0),
        # Simple subtraction
        (["10", "5", "-"], 5),
    ],
)
def test_eval_rpn(tokens, expected):
    assert eval_rpn(tokens) == expected


def test_single_token():
    """Test case for a single number with no operators."""
    assert eval_rpn(["42"]) == 42

File: stacks/evaluate_reverse_polish_notation/eval_polish_notation_test.go
package rpn

import (
	"testing"
)

func TestEvalRPN(t *testing.T) {
	tests := []struct {
		name     string
		tokens   []string
		expected int
	}{
		{
			name:     "Basic addition and multiplication",
			tokens:   []string{"2", "1", "+", "3", "*"},
			expected: 9, // ((2 + 1) * 3)
		},
		{
			name:     "Substraction and division",
			tokens:   []string{"4", "13", "5", "/", "+"},
			expected: 6, // (4 + (13 / 5)) -> 4 + 2
		},
		{
			name:     "Complex expression with negative results",
			tokens:   []string{"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"},
			expected: 22,
		},
		{
			name:     "Truncation toward zero (positive)",
			tokens:   []string{"1", "2", "/"},
			expected: 0,
		},
		{
			name:     "Truncation toward zero (negative)",
			tokens:   []string{"-1", "2", "/"},
			expected: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := EvalRPN(tt.tokens)
			if result != tt.expected {
				t.Errorf("EvalRPN(%v) = %d; want %d", tt.tokens, result, tt.expected)
			}
		})
	}
}
